.gitlab-ci.yml

```diff
@@ -48,19 +48,28 @@ lint_test_scripts:
 .build_environment_common: &build_environment_common
   <<: *parameters_common
   image: registry.hub.docker.com/library/docker:stable
+  before_script:
+    - |
+      apk update && apk add git
 
 prepare_build_environment:
   <<: *build_environment_common
-  only:
-    <<: *changes_docker
   stage: prepare
   script:
-    - docker build --rm -t "${CI_COMMIT_SHA}:${CI_PIPELINE_ID}" .
     - |
+      if git whatchanged --name-only --pretty="" master...HEAD | grep "Dockerfile\|.dockerignore"; then
+        echo "Dockerfile changes detected..."
+        docker build --rm -t "${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHA}" .
         docker run \
         --rm \
         "$CI_COMMIT_SHA:$CI_PIPELINE_ID" \
         /test/buildenv_check.sh
+      else
+        echo "No Dockerfile changes..."
+        docker login -u gitlab-ci-token -p "${CI_JOB_TOKEN}" "${CI_REGISTRY}"
+        docker tag "${CI_REGISTRY_IMAGE}:latest" "${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHA}"
+        docker push "${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHA}"
+      fi
```

Comment:
When there are changes, we make an image and tag is as the commit SHA.

When there are no changes, we tag :latest as the commit SHA.